compiler/backend/jsgen.nim

```diff
@@ -1377,53 +1370,47 @@ proc genSymAddr(p: PProc, n: PNode, typ: PType, r: var TCompRes) =
   else: internalError(p.config, n.info, $("genAddr: 2", s.kind))
 
 proc genAddr(p: PProc, n: PNode, r: var TCompRes) =
-  if n.kind == nkSym:
-    genSymAddr(p, n, nil, r)
-  else:
-    case n[0].kind
-    of nkSym:
-      genSymAddr(p, n[0], n.typ, r)
-    of nkCheckedFieldExpr:
-      genCheckedFieldOp(p, n[0], n.typ, r)
-    of nkDotExpr:
-      if mapType(p, n.typ) == etyBaseIndex:
-        genFieldAddr(p, n[0], r)
-      else:
-        genFieldAccess(p, n[0], r)
-    of nkBracketExpr:
-      var ty = skipTypes(n[0].typ, abstractVarRange)
-      if ty.kind in MappedToObject:
-        gen(p, n[0], r)
-      else:
-        let kindOfIndexedExpr = skipTypes(n[0][0].typ, abstractVarRange).kind
-        case kindOfIndexedExpr
-        of tyArray, tyOpenArray, tySequence, tyString, tyCstring, tyVarargs:
-          genArrayAddr(p, n[0], r)
-        of tyTuple:
-          genFieldAddr(p, n[0], r)
-        of tyGenericBody:
-          genAddr(p, n[^1], r)
-        else: internalError(p.config, n[0].info, "expr(nkBracketExpr, " & $kindOfIndexedExpr & ')')
-    of nkObjDownConv:
-      gen(p, n[0], r)
-    of nkHiddenDeref:
-      gen(p, n[0], r)
-    of nkHiddenAddr:
-      gen(p, n[0], r)
-    of nkConv:
-      genAddr(p, n[0], r)
-    of nkStmtListExpr:
-      p.config.internalAssert(n.len == 1, n[0].info, "genAddr for complex nkStmtListExpr")
-      gen(p, n[0], r)
-    of nkCallKinds:
-      if n[0].typ.kind == tyOpenArray:
-        # 'var openArray' for instance produces an 'addr' but this is harmless:
-        # namely toOpenArray(a, 1, 3)
-        gen(p, n[0], r)
-      else:
-        internalError(p.config, n[0].info, "genAddr: " & $n[0].kind)
+  ## Dispatches to the appropriate procedure for generating the address-of
+  ## operation based on the kind of node `n`
+  case n.kind
+  of nkSym:
+    genSymAddr(p, n, r)
+  of nkCheckedFieldExpr:
+    genCheckedFieldOp(p, n, takeAddr=true, r)
+  of nkDotExpr:
+    genFieldAddr(p, n, r)
+  of nkBracketExpr:
```

Comment:
Thinking or loud: This bit really highlights how the AST is just off, it really should be lowered/specialized further. The type based dispatch within the branch seems off.